### PR TITLE
PKCS7: added encryption with AES-256-CBC

### PR DESCRIPTION
**Pull request by nitneuqr**
A few things to check: 
- Should we name the parameters `content_algorithm` instead of just `algorihm`
- I also have a version with `mode` included (as we might need it on next S/MIME versions), should I include it now?
- Maybe go even simpler with an enum (AES128 & AES256) ?

### CODE REVIEW

File: docs/hazmat/primitives/asymmetric/serialization.rst [hunk 1]
```diff
@@ -1284,6 +1287,13 @@ contain certificates, CRLs, and much more. PKCS7 files commonly have a ``p7b``,
         :param data: The data to be encrypted.
         :type data: :term:`bytes-like`
 
+    .. method:: set_algorithm(algorithm)
+
+        Sets the content encryption algorithm to use. Only AES-128 and AES-256 are supported.
+        
+        :type algorithm: A :class:`~cryptography.hazmat.primitives.ciphers.algorithms.AES`
+            with a key size of 128 or 256 bits.
```
Comment (alex) [inline docs/hazmat/primitives/asymmetric/serialization.rst:1295]: I find this documentation a bit difficult to read.

File: src/cryptography/hazmat/primitives/serialization/pkcs7.py [hunk 2]
```diff
@@ -221,6 +234,19 @@ def add_recipient(
                 *self._recipients,
                 certificate,
             ],
+            _algorithm=self._algorithm,
+        )
+
+    def set_algorithm(
+        self, algorithm: type[CipherAlgorithm]
+    ) -> PKCS7EnvelopeBuilder:
```
Comment (alex) [inline src/cryptography/hazmat/primitives/serialization/pkcs7.py:242]: Generally we make it an error to call something multiple times.

File: src/rust/src/pkcs7.rs [hunk 3]
```diff
@@ -95,14 +95,19 @@ fn encrypt_and_serialize<'p>(
         smime_canonicalize(raw_data.as_bytes(), text_mode).0
     };
 
-    // The message is encrypted with AES-128-CBC, which the S/MIME v3.2 RFC
-    // specifies as MUST support (https://datatracker.ietf.org/doc/html/rfc5751#section-2.7)
-    let key = types::OS_URANDOM.get(py)?.call1((16,))?;
-    let aes128_algorithm = types::AES128.get(py)?.call1((&key,))?;
+    // Get the algorithm
+    let algorithm_type = builder.getattr(pyo3::intern!(py, "_algorithm"))?;
+    let key_size = algorithm_type.getattr(pyo3::intern!(py, "key_size"))?;
```
Comment (alex) [inline src/rust/src/pkcs7.rs:100]: This is wrong for many `type[CipherAlgorithm]`, it really onyl works with AES128 and AES256.

Comment (nitneuqr) [reply src/rust/src/pkcs7.rs:100]: That's correct, blocking it in the Python side!

File: docs/hazmat/primitives/asymmetric/serialization.rst (hunk 1 above)
Comment (nitneuqr) [reply docs/hazmat/primitives/asymmetric/serialization.rst]: Tried to simplify it, please let me know!

File: src/cryptography/hazmat/primitives/serialization/pkcs7.py (hunk 2 above)
Comment (nitneuqr) [reply src/cryptography/hazmat/primitives/serialization/pkcs7.py:241]: Done!

File: src/rust/src/pkcs7.rs (hunk 3 above)
Comment (nitneuqr) [reply src/rust/src/pkcs7.rs]: Simplified the way it works to only handle the 2 algorithm types.

**Comment by nitneuqr** [comment]
Made some changes to adapt, let me know what you think @alex !

File: src/cryptography/hazmat/primitives/serialization/pkcs7.py [hunk 4]
```diff
@@ -232,6 +264,13 @@ def encrypt(
             raise ValueError("Must have at least one recipient")
         if self._data is None:
             raise ValueError("You must add data to encrypt")
+
+        # The default content encryption algorithm is AES-128, which the S/MIME
+        # v3.2 RFC specifies as MUST support (https://datatracker.ietf.org/doc/html/rfc5751#section-2.7)
+        self._content_encryption_algorithm = (
+            self._content_encryption_algorithm or algorithms.AES128
+        )
```
Comment (alex) [inline src/cryptography/hazmat/primitives/serialization/pkcs7.py:272]: Don't mutate self here, just use a local. You'll get wrong behavior on a reused builder.

Comment (nitneuqr) [reply src/cryptography/hazmat/primitives/serialization/pkcs7.py:272]: Good catch! I'll need to modify the rust code accordingly

Comment (nitneuqr) [reply src/cryptography/hazmat/primitives/serialization/pkcs7.py]: Done!

**Comment by alex** [comment]
Looks like there's a merge conflict in the changelog. can you rebase/merge main before I review?

**Comment by nitneuqr** [comment]
> Looks like there's a merge conflict in the changelog. can you rebase/merge main before I review?

Yup, done!

**Review by alex — APPROVED** [review]

**Pull request merged**